source/Halibut/Transport/Protocol/MessageSerializer.cs

```diff
@@ -61,17 +67,28 @@ public void WriteMessage<T>(Stream stream, T message)
                 createSerializer().Serialize(bson, new MessageEnvelope<object> { Message = message });
             }
 
-            observer.MessageWritten(compressedByteCountingStream.BytesWritten);
+            observer.MessageWritten(compressedByteCountingStream.BytesWritten, 0);
         }
 
         public async Task WriteMessageAsync<T>(Stream stream, T message, CancellationToken cancellationToken)
         {
-            // TODO - ASYNC ME UP!
-            await Task.CompletedTask;
+            using var compressedByteCountingStream = new ByteCountingStream(stream, OnDispose.LeaveInputStreamOpen);
+            using var compressedInMemoryBuffer = new WriteIntoMemoryBufferStream(compressedByteCountingStream, writeIntoMemoryLimitBytes, OnDispose.LeaveInputStreamOpen);
```

Comment:
Could this just be a [BufferedStream](https://learn.microsoft.com/en-us/dotnet/api/system.io.bufferedstream?view=net-7.0)?

Reply:
It seems like it could.

Looking through the BufferedStream code, I think it would work, as it tends to write to the memory buffer until it reaches the buffer size, then it writes out to the underlying stream. 

But then once the buffer is written to the stream when buffer is full, it then clears and repeats. I'm not sure if this is what we want. I don't think it really matters though.

The BufferedStream does not expose how much was written through the buffer. So to be able to support `compressedInMemoryBuffer.BytesWrittenIntoMemory`, we'd have to calculate that by using the min of `compressedByteCountingStream.BytesWritten` and the cap.

The above is quite minor. The only thing making me pause on cutting over is that we lose control of how this works, and can't tweak as easily in future.

But if you think it worth doing, I can do it 🙂

Reply:
I guess the difference is once the buffer is exceeded the current stream wont do any more buffering where BufferedStream will.

Reply:
For now, I'm inclined to leave it as is, and then we can play with it in subsequent PRs